Feature/doo hickey.v4

jasonplatte opened the pull request:
PTAL

jasonplatte committed:
Added Group Command and changed FX to SRX

Changed all places where it said TalonFX to TalonSRX and made a group command that engages the doohickey, waits 3 seconds, then spins a distance, and retracts. Also added a stop to "Spin" command.

jasonplatte committed:
Added Components

Support the component interface for DooHickey

jasonplatte committed:
Added Array of Structs 

For loop logic is commented out because it has an error, could also be made more efficient.

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -131,6 +131,98 @@ void DooHickey::Periodic() {
       IsRotationControlFinished = true;
     }
 
+    struct colorstuff 
```
Let's spend some time on this on Monday (or when we have the shooter working). I have some recommendations for how to make some changes so that this is easier to use. We also need to move this to a command instead of in periodic.

duane701 commented on src/main/include/RobotContainer.h:
```diff
@@ -97,10 +97,13 @@ class RobotContainer {
-                                    kDoohickeyPushyThangSolenoidReverse};
-  DooHickey mDooHickey{"DooHickey", mDooHickeySpinner, mUpPushyThang};
+  //WPI_TalonSRX mSpinner{kDoohickey};
+  //frc::DoubleSolenoid mUpPushyThang{kDoohickeyPushyThangSolenoidFoward, kDoohickeyPushyThangSolenoidReverse};
```
Let's delete these comments

duane701 commented on src/main/include/RobotContainer.h:
```diff
@@ -97,10 +97,13 @@ class RobotContainer {
+  //frc::DoubleSolenoid mUpPushyThang{kDoohickeyPushyThangSolenoidFoward, kDoohickeyPushyThangSolenoidReverse};
+  DooHickey::Components mDooHickeyComponents{
+    WPI_TalonSRX{11},
+    frc::DoubleSolenoid{0,7}
```
There are symbols for the doohicky component ids

```suggestion
    frc::DoubleSolenoid{kDoohickeyPushyThangSolenoidFoward, kDoohickeyPushyThangSolenoidReverse}
```

duane701 commented on src/main/include/RobotContainer.h:
```diff
@@ -97,10 +97,13 @@ class RobotContainer {
+  //WPI_TalonSRX mSpinner{kDoohickey};
+  //frc::DoubleSolenoid mUpPushyThang{kDoohickeyPushyThangSolenoidFoward, kDoohickeyPushyThangSolenoidReverse};
+  DooHickey::Components mDooHickeyComponents{
+    WPI_TalonSRX{11},
```
```suggestion
    WPI_TalonSRX{kDoohickey},
```

duane701 commented on src/main/cpp/commands/RotateHickey.cpp:
```diff
@@ -0,0 +1,28 @@
+ {
+  constexpr double radiusCW = 16; //16" radius of Control panel 
+  constexpr double radiusDW = 1.5;  //1.5" radius of DooHickey wheel (3" diameter)
+  constexpr double TargetPos = (4*(radiusCW/radiusDW)) * 2048; 
```
The other magic numbers would be useful as well.

``` c++
constexpr double kNumberOfRotations{4};
constexpr double kTicksPerRotation{2048};
```

duane701 commented:
FTP

jasonplatte committed:
Added constants

jasonplatte committed:
Added functions and Array of structs inside of Array of structs

Command coming up in the next episode.

jasonplatte committed:
Made Command that uses array of structs

Builds in 3 minutes :.(

jasonplatte committed:
Removed some Magic Numbers and added Dashboard button

Made variables for some numbers to make it easier to read and made a smart dashboard button for the position control command.

duane701 commented on src/main/cpp/RobotContainer.cpp:
```diff
@@ -66,8 +69,8 @@ RobotContainer::RobotContainer()
   
-  frc::SmartDashboard::PutData("Spin 600 RPM", new Spin(mDooHickey, 0.1));
-  frc::SmartDashboard::PutData("Spin 6000 RPM", new Spin(mDooHickey, 0.9404));
+  frc::SmartDashboard::PutData("Spin 58 RPM", new Spin(mDooHickey, 18730.0)); //RPM for 775Pro = 58 RPM on wheel.
```
Let's include the math for how `18730` is determined.

duane701 replied:
What is this number? Seems really odd.

duane701 replied:
Seems like `18730` should be `58`.

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
+
+  if(abs(mComponents.spinner.GetActiveTrajectoryPosition() - TargetPos) < kErrorThreshold)
+  {
+    kWithinThresholdLoops++;
```
`kWithinThresholdLoops` should be `mWithinThresholdLoops`

duane701 reviewed (changes requested):
These are the simple changes. 

I'm looking at the code for more comments.

duane701 commented on src/main/cpp/RobotContainer.cpp:
```diff
@@ -43,9 +44,11 @@ RobotContainer::RobotContainer()
+  constexpr double kTicksPerRotation{2048};
+  constexpr double radiusCW{16}; //16" radius of Control panel 
+  constexpr double radiusDW{1.5};  //1.5" radius of DooHickey wheel (3" diameter)
+  constexpr double TargetPos = (kNumberOfRotations*(radiusCW/radiusDW)) * kTicksPerRotation;
```
Seems like all of these values should be in a file more closed related to the DooHickey.

duane701 replied:
After practice

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -10,22 +10,23 @@
-  :mMotorSpeed{0}
+DooHickey::DooHickey(const wpi::Twine& name, Components& components)
+: mComponents(components)
+  ,mMotorSpeed{0}
```
Remove `mMotorSpeed` just call the function with the value.

duane701 replied:
After practice

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -65,7 +66,10 @@ void DooHickey::Init() {
+  constexpr double kSecondsTo100ms{10};
+  double setSpeedTP100ms = speed * kTicksPerRotation / kSecondsPerMinute / kSecondsTo100ms;
   mMotorSpeed = setSpeedTP100ms;
   UpdateSpeed();
```
UpdateSpeed should take the speed

```suggestion
  UpdateSpeed(motorSpeed);
```

duane701 replied:
After practice

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
 }
 
 void DooHickey::UpdateSpeed() {
```
```suggestion
void DooHickey::UpdateSpeed(double speedTicks) {
```

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
 
 void DooHickey::UpdateSpeed() {
-  mSpinner.Set(ControlMode::Velocity, mMotorSpeed);
+  mComponents.spinner.Set(ControlMode::Velocity, mMotorSpeed);
```
```suggestion
  mComponents.spinner.Set(ControlMode::Velocity, speedTicks);
```

jasonplatte committed:
changed number to bool

duane701 commented on src/main/include/subsystems/DooHickey.h:
```diff
@@ -56,11 +87,16 @@ class DooHickey : public frc2::SubsystemBase {
   void UpdateSpeed();
```
```suggestion
  void UpdateSpeed(double speedTicks);
```

duane701 replied:
When would we use `UpdateSpeed`?

MotionMagic for spin a distance, PercentVBus for move to the right color.

duane701 replied:
I see we war are using Velocity mode for moving. That ok, but may be overkill.

duane701 commented on src/main/cpp/RobotContainer.cpp:
```diff
@@ -66,8 +69,8 @@ RobotContainer::RobotContainer()
-  frc::SmartDashboard::PutData("Spin 600 RPM", new Spin(mDooHickey, 0.1));
-  frc::SmartDashboard::PutData("Spin 6000 RPM", new Spin(mDooHickey, 0.9404));
+  frc::SmartDashboard::PutData("Spin 58 RPM", new Spin(mDooHickey, 18730.0)); //RPM for 775Pro = 58 RPM on wheel.
+  frc::SmartDashboard::PutData("Position Control", new HickeyPosControl(mDooHickey, 9365.0));
```
I think this is also RPM. This seems like not the right value.

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
+  frc::Color matchedColor = m_colorMatcher.MatchClosestColor(detectedColor, confidence);
+  frc::Color assignedColor;
+
+    colorstuff Selector1[] = 
```
Move `Selector1`  to the anonymous namespace.

This works here for now, so let's move `Selector1` later.

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
+DooHickey::stoppingColor DooHickey::StopHereColor()
+{
+  frc::Color matchedColor = m_colorMatcher.MatchClosestColor(detectedColor, confidence);
+  frc::Color assignedColor;
```
`assignedColor` needs a value.

duane701 replied:
See comment below in `Periodic`

Declare `assignedColor` as a member variable in the header file, with the name `mAssignedColor`. Don't declare here at all. Just use `mAssignedColor`.

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -103,14 +174,10 @@ void DooHickey::Periodic() {
+
+  frc::Color matchedColor = m_colorMatcher.MatchClosestColor(detectedColor, confidence);
+  frc::Color detectedColor = m_colorSensor.GetColor();
+  frc::Color assignedColor;
```
Move `assignedColor` from a local to a member variable, `mAssignedColor`. Declare `mAssignedColor` in the header for for the DooHickey class.

The uses from `assignedColor` as a local to `mAssignedColor` as a member variable.

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -10,22 +10,23 @@
+DooHickey::DooHickey(const wpi::Twine& name, Components& components)
+: mComponents(components)
+  ,mMotorSpeed{0}
   ,mTargetPos{0}
```
Just like `mMotorSpeed`, `mTargetPos` can be a local instead of a member variable.

duane701 replied:
After practice

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
+  mComponents.spinner.Set(ControlMode::Velocity, mMotorSpeed);
 }
 
 void DooHickey::UpdatePos() {
```
```suggestion
void DooHickey::UpdatePos(double targetPos) {
```

duane701 replied:
After practice

duane701 commented on src/main/cpp/subsystems/DooHickey.cpp:
```diff
@@ -76,19 +80,86 @@ void DooHickey::SetHickeyPosition(double position){
 
 void DooHickey::UpdatePos() {
-  mSpinner.Set(ControlMode::MotionMagic, mTargetPos);
+  mComponents.spinner.Set(ControlMode::MotionMagic, mTargetPos);
```
```suggestion
  mComponents.spinner.Set(ControlMode::MotionMagic, targetPos);
```

duane701 replied:
After practice

duane701 commented on src/main/include/subsystems/DooHickey.h:
```diff
@@ -29,22 +35,47 @@ class DooHickey : public frc2::SubsystemBase {
 
   double mMotorSpeed;
   double mTargetPos; 
```
Remove `mTargetPos` as a class member. Make only a local or parameter. (Same as `mMotorSpeed`)

duane701 replied:
After practice

duane701 commented on src/main/include/subsystems/DooHickey.h:
```diff
@@ -56,11 +87,16 @@ class DooHickey : public frc2::SubsystemBase {
   void UpdateSpeed();
   void UpdatePos();
```
```suggestion
  void UpdatePos(double positionTicks);
```

duane701 replied:
After practice

duane701 reviewed (changes requested):
Lots of changes.

I will comment on the required changes.

jasonplatte committed:
Added member variables

duane701 committed:
DooHickey: Syntax fixes

jasonplatte committed:
Renamed button

"Spin distance" was confusing. Renamed to Rotation Control.

bookslifestyle committed:
Worked on DooHickey Issues

Changed SetHickeyPos to a command instead of an instant command and removed member variables. Tried to move Selector array into namespace but ran into some problems. Confused on how to do matched color and assigned color functions.

toxicoreos committed:
Worked on Issues part 2

Moved selector 1 to the namespace. Made gamedata function. Made matchedColor function.
